### ISSUE TITLE
Performance issue with predict one

### ISSUE THREAD

**Issue by sanyam-cashify** [online-ml/river#811]
## Versions

**`river` version**: 0.8.0

**Python version**: 3.6

## Describe your task

pipe_db = Pipeline(('vectorizer', TFIDF(lowercase=True, ngram_range=(3,4))),('db_scan',DBSTREAM(clustering_threshold=0.3, minimum_weight=5)))

The training time is fine ,but when i'm obtaining clusters from the same by using predict_one for each text entity, it's taking too long, the time taken increases with O(n). Here's the code->

data["cluster_labels"] = data["sq_address"].progress_apply(lambda x: model.predict_one(x))

## What kind of performance are you expecting?
Reduction in prediction time.

**Comment by MaxHalford** [comment]
Hello there :). Could you please provide a [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example)?

**Comment by sanyam-cashify** [comment]
class MODEL_DEPLOYMENT:
    def __init__(self):
        self.data = pd.read_csv(f"{PATH}datasets/Xpress orders 12dec - Sheet 1 (1).csv")

    def get_pipeline(self) -> Pipeline:
        pipe_db = Pipeline(
            ("vectorizer", TFIDF(lowercase=True, ngram_range=(3, 4))),
            ("db_scan", DBSTREAM(clustering_threshold=0.3, minimum_weight=5)),
        )
        return pipe_db

    def db_scan(self, ratio: int = 1) -> Pipeline:
        data = self.data
        data = data[["sq_address"]]
        data = data[: int(len(data) * ratio)]
        data = data.to_records(index=False)
        pipeline = self.get_pipeline()
        for text in tqdm(data):
            pipeline = pipeline.learn_one(text[0])
        return pipeline

    def gen_pred(self, ratio: float = 1.0) -> pd.DataFrame:
        data = self.data.sample(5000)
        data = data[["sq_address"]]
        # data = data.to_records(index = False)
        model = self.db_scan(ratio=ratio)
        # predL = []
        # dfPred = data["sq_address"]
        # testL = []
        # testT = tuple(data["sq_address"].values.tolist())
        # testL.append(testT)
        # print(testL[:5])
        # predL = model.predict_many(data)
        # print(predL)

        # for text in tqdm(data["sq_address"].values.tolist()):
        #     pred = model.predict_one(text)
        #     predL.append(pred)
        data["cluster_labels"] = data["sq_address"].progress_apply(
            lambda x: model.predict_one(x),
        )
        return data


It takes about 25:09 minutes for 5000 datapoints.

**Comment by MaxHalford** [comment]
Hey again. Sorry for the late answer. Could you provide you're using? You can send it to `maxhalford25@gmail.com` if you're not comfortable sharing it here.

**Comment by sanyam-cashify** [comment]
Using what?

**Comment by MaxHalford** [comment]
Sorry, I meant "could you provide the dataset you're using".

**Comment by srnthsrdhrn** [comment]
I found a small workaround to speed up the prediction. Not sure if it has some side effects. Sharing it here.

`self.denstream.predict_one(river_utils.numpy2dict(embeddings[0])) # This is done once so that the cluster initialization happens
  cluster_ids = []
  for x,_ in tqdm(stream.iter_array(embeddings),total=len(embeddings)):
      out = self.denstream._get_closest_cluster_key(x, self.denstream.clusters)`

When I also faced this issue of too much time for prediction, I went through the code and found that the reason predict_one takes time is because, whenever we ask for a prediction, the neighbors and clusters are computed.

This makes sense for a streaming algorithm since we expect the cluster to keep evolving and hence we need to always give the most recent answer.

But my use case was that 
1. I learn for sometime, then
2. I request prediction for a bunch of points and
3.  then I learn again. 

So when I am requesting the points, there is no new data coming into the model. So it felt unnecessary to rebuild the cluster for every prediction. 
So I came up with my own workaround here. 

Maybe @MaxHalford  can verify if this makes sense or if I am doing something wrong.
